At first send open the connection if not already opened.

toonknapen opened the pull request:
This makes it much easier to instantiate the TCPLogstashHandler using the logging.config and have it immediately ready for use

<!-- Thank you for your contribution! -->

## What do these changes do?

This allows to create the handler using the factory method of logging.config.dictConfig and have the logger directly available afterwards.

## Are there changes in behavior for the user?

No

## Related issue number

Not to my knowledge. 

## Checklist

- [ ] I think the code is well written
- [ ] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [ ] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` (e.g. `588.bugfix`)
  * if you don't have an `issue_id` change it to the pr id after creating the PR
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: `Fix issue with non-ascii contents in doctest text files.`

hellysmile commented:
Hey @toonknapen, thanks for the patch! 
Can You please add test cases?
Small example to Readme how to use handler would be great 
Also adding setup.py, should be done in separate pull request and requires additional discussion.

asvetlov commented:
I believe `setup.py` is a wrong direction, `flit.ini` should be converted into `pyproject.toml` instead.
Anyway, it is a subject for separate PR.

toonknapen commented:
Will add test-cases and example on how this improves the usability and will do a new PR for that.
Indeed setup.py should not have been included but I had added it to be able to `pip install -e` the library to test the modification in another project.

toonknapen commented:
Before doing a new PR, please have a look at https://github.com/toonknapen/aiologstash/commit/7f033f889423648134e160610e57bc26af15c668 where I demonstrate the issue in the test 'test_simple_using_config'.

The test is currently not very clean but it points out 2 problems:

1) the 'level' parameter of the TCPLogstashHandler.__init__
When there is a 'level' attribute in the dictionary passed to the logging.config.dictConfig, it will not be passed to the constructor of the handler and thus the instantiation of the TCPLogstashHandler will fail.

To bypass this problem I added another handler that does not require a level parameter to be passed. This handler should be dropped of course but just added it temporarily to show the issue.

2) the TCPLogstashHandler._writer is not instantiated
When running the test I added, you will see that the  TCPLogstashHandler._send will be called while the TCPLogstashHandler._writer is still None and thus will fail.
The test will never finish as the fake server will thus never receive anything.
Again the idea is to leave it like that but just wanted to point out why I proposed the fix in my initial PR.

If there is agreement on a solution, I propose I will do a new PR.

asvetlov commented:
Hold on.
The handler should be created by `await create_tcp_handler()`.
Explicit instantiation of `TCPLogstashHandler` instance is forbidden by design.
If you instantiate the library properly everything should work just fine.

toonknapen commented:
But that thus means that I can not just read my logging.config from a file?

For instance for a dev, preprod and prod environment I have exactly the same code but just a different logging.config file. But thus I need to install the handlers and formatters just from a config file. So how can I achieve that?

asvetlov commented:
The problem is: in asyncio world, all objects should live inside a loop.
Any object is instantiated when the loop exists and running, the object should be destroyed before the loop closing.
The rule is not super strict, but the violation leads to hard-to-debug problems.
Theoretically, you can design a system in a way where the async handler is instantiated by file configuration. But the loader should be disabled, switching it on requires an explicit operation.

toonknapen commented:
I'm calling logging.config.dictConfig only after the loop is created so that is fine right?

asvetlov commented:
Do you call it from async function?

toonknapen commented:
Yes, see https://github.com/toonknapen/aiologstash/blob/7f033f889423648134e160610e57bc26af15c668/tests/test_simple.py#L29 which calls https://github.com/toonknapen/aiologstash/blob/7f033f889423648134e160610e57bc26af15c668/tests/conftest.py#L144 which on its turn calls `logging.config.dictConfig`

asvetlov commented:
Well, in this particular case your patch might work.
To make it more robust please raise an exception if the logstash handler is instantiated without running loop, like async-timeout does now: https://github.com/aio-libs/async-timeout/blob/master/async_timeout/__init__.py#L212-L219

toonknapen commented:
Good idea. OK will do and will soon provide an update and will do a PR

toonknapen commented:
Just to be clear: where do you prefer me to add the check if the loop is running. In a separate handler (derived from TCPLogstashHandler) or can I add it directly to TCPLogstashHandler ?

Taking into account to that currently TCPLogstashHandler can not be instantiated through logging.config.dictConfig because of the signature of the __init__ as it requires a 'level' argument which is interpreted by the dictConfig itself (and thus not passed anymore to the constructor of the handler)

So I either adapt the TCPLogstashHandler or provide a handler derived from TCPLogstashHandler. The latter is then allowed to be instantiated by the user directly or by something like logging.config.dictConfig (while the former isn't)

asvetlov commented:
I'm very sorry.
I thought I've answered already.

I think modifying `TCPLogstashHandler.__init__` signature is fine, the constructor is private and was never used by user-side code.
`logging.Handler()` accepts `level` argument with the `logging.NOTSET` default.
I think we can keep it.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/aio-libs/aiologstash?pullRequest=107) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/aio-libs/aiologstash?pullRequest=107) before we can accept your contribution.<br/><hr/>**Toon Knapen** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/aio-libs/aiologstash?pullRequest=107) it.</sub>
